### PR TITLE
[inductor] `_sparse_semi_structured_linear` fallback - no meta registration; not on testing path

### PR DESCRIPTION
**Pull request by jon-chuang**
Test was wrong in original PR and merged changes were never tested. Further, the sparse op was never actually compiled due to missing `fullgraph=True` and missing meta registration. 

When meta is added as per this PR, it gives wrong answers when input needs to be padded and when input needs to be reshaped.

Is this something to do with the generated inductor code for:
```
 constant_pad_nd: "f16[32, 128]" = torch.ops.aten.constant_pad_nd.default(primals_3, [0, 0, 0, 31], 0.0)
...
slice_1: "f16[1, 128]" = torch.ops.aten.slice.Tensor(_sparse_semi_structured_linear, 0, 0, 1);  _sparse_semi_structured_linear = None
```
and

```
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         mul: "Sym(s0*s1)" = primals_4 * primals_5
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         view: "f16[s0*s1, 128]" = torch.ops.aten.view.default(primals_6, [mul, 128]);  primals_6 = mul = None
...
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         view_1: "f16[s0, s1, 128]" = torch.ops.aten.view.default(slice_1, [primals_4, primals_5, 128]);  slice_1 = None
```

Failing graphs:
Padded:
```
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]  ===== Forward graph 5 =====
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.66 class GraphModule(torch.nn.Module):
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, primals_1: "f16[128, 64]", primals_2: "i16[128, 8]", primals_3: "f16[1, 128]"):
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/jonch/Desktop/Programming/mlsys/pytorch/test/test_sparse_semi_structured.py:145, code: x = self.linear(x)
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         constant_pad_nd: "f16[32, 128]" = torch.ops.aten.constant_pad_nd.default(primals_3, [0, 0, 0, 31], 0.0)
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         _sparse_semi_structured_linear: "f16[32, 128]" = torch.ops.aten._sparse_semi_structured_linear.default(constant_pad_nd, primals_1, primals_2);  constant_pad_nd = primals_1 = primals_2 = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         slice_1: "f16[1, 128]" = torch.ops.aten.slice.Tensor(_sparse_semi_structured_linear, 0, 0, 1);  _sparse_semi_structured_linear = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         slice_2: "f16[1, 128]" = torch.ops.aten.slice.Tensor(slice_1, 1, 0, 9223372036854775807);  slice_1 = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/jonch/Desktop/Programming/mlsys/pytorch/test/test_sparse_semi_structured.py:147, code: return torch.nn.functional.relu(x)
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         relu: "f16[1, 128]" = torch.ops.aten.relu.default(slice_2);  slice_2 = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         alias: "f16[1, 128]" = torch.ops.aten.alias.default(relu)
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         alias_1: "f16[1, 128]" = torch.ops.aten.alias.default(alias);  alias = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         le: "b8[1, 128]" = torch.ops.aten.le.Scalar(alias_1, 0);  alias_1 = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/jonch/Desktop/Programming/mlsys/pytorch/test/test_sparse_semi_structured.py:145, code: x = self.linear(x)
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         permute: "f16[128, 1]" = torch.ops.aten.permute.default(primals_3, [1, 0]);  primals_3 = None
[2023-11-23 13:59:51,102] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return [relu, le, permute]

```

Reshape:

```
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.69 class GraphModule(torch.nn.Module):
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, primals_1: "f16[128, 64]", primals_2: "i16[128, 8]", primals_3: "f16[128]", primals_4: "Sym(s0)", primals_5: "Sym(s1)", primals_6: "f16[s0, s1, 128]"):
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/jonch/Desktop/Programming/mlsys/pytorch/test/test_sparse_semi_structured.py:145, code: x = self.linear(x)
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         mul: "Sym(s0*s1)" = primals_4 * primals_5
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         view: "f16[s0*s1, 128]" = torch.ops.aten.view.default(primals_6, [mul, 128]);  primals_6 = mul = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         _sparse_semi_structured_linear: "f16[s0*s1, 128]" = torch.ops.aten._sparse_semi_structured_linear.default(view, primals_1, primals_2, bias = primals_3);  primals_1 = primals_2 = primals_3 = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         slice_1: "f16[s0*s1, 128]" = torch.ops.aten.slice.Tensor(_sparse_semi_structured_linear, 1, 0, 9223372036854775807);  _sparse_semi_structured_linear = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         view_1: "f16[s0, s1, 128]" = torch.ops.aten.view.default(slice_1, [primals_4, primals_5, 128]);  slice_1 = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/jonch/Desktop/Programming/mlsys/pytorch/test/test_sparse_semi_structured.py:147, code: return torch.nn.functional.relu(x)
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         relu: "f16[s0, s1, 128]" = torch.ops.aten.relu.default(view_1);  view_1 = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         alias: "f16[s0, s1, 128]" = torch.ops.aten.alias.default(relu)
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         alias_1: "f16[s0, s1, 128]" = torch.ops.aten.alias.default(alias);  alias = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         le: "b8[s0, s1, 128]" = torch.ops.aten.le.Scalar(alias_1, 0);  alias_1 = None
[2023-11-23 14:01:03,463] [0/2] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return [relu, view, le, primals_4, primals_5]

```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by jon-chuang** [comment]
@jcaip  your PR https://github.com/pytorch/pytorch/pull/111049 was never tested and is wrong

Could you please look into fixing?

**Comment by jcaip** [comment]
Thanks for flagging this, cc @alexsamardzic 

Would you happen to know if there are any special cases in the case of padded/reshaped inputs passed to  _sparse_semi_structured_linear? 

I don't see an issue for the (1, 128) and (64, 128, 128) test cases for _cslt_sparse_mm when I add in the meta registrations used here: https://github.com/pytorch/pytorch/pull/114370.

**Comment by alexsamardzic** [comment]
I confirm there is an issue there with this PR applied, and when CUTLASS backend used - the simpler reproducer is as follows:

<details open>
<summary>Reproducer script</summary>

```Python
import torch

from torch.sparse.semi_structured import (
    SparseSemiStructuredTensor,
    to_sparse_semi_structured,
)

@torch.compile(backend="inductor", fullgraph=True)
def my_linear(input, weight):
    return torch.nn.functional.linear(input, weight)

SparseSemiStructuredTensor._FORCE_CUTLASS = True

m, n, k = 1, 32, 64
dtype = torch.half
device = "cuda"

torch.manual_seed(0)

input = torch.rand((m, k), dtype=dtype, device=device)
weight = torch.rand((n, k), dtype=dtype, device=device)

mask = torch.Tensor([1, 0, 0, 1]).to(dtype).to(device).tile((n, k // 4))

dense_weight = weight * mask
dense_result = torch.nn.functional.linear(input, dense_weight)

sparse_weight = to_sparse_semi_structured(dense_weight)
sparse_result = my_linear(input, sparse_weight)

assert torch.allclose(dense_result, sparse_result, rtol=1e-3, atol=1e-3)
```
</details>

When I comment out `@torch.compile` line in the script above, the results for both dense and sparse case are the same, but with this line, these differ.  However, I added printouts in the C++ code in `_sparse_semi_structured_linear()` function, that is calculating the linear operator, and in both cases inputs and outputs are the same.   So after examining the generated code, I strongly suspect that the problem is in the line:
```
return (reinterpret_tensor(buf2, (1, 32), (32, 1), 0), )
```
that gets generated below `aten._sparse_semi_structured_linear()` call.  So I'm not sure where the fix is to be made?

(Above was all for the case of `(1, 64)` input.  For the other problematic input size, with batch dimension, it seems to be the same kind of problem, however in this case not because of padding, but instead because the CUTLASS version is internally squashing batch dimensions of input tensor together with row dimension - but it is also internally properly setting back the dimensions of the result vector, so the tensor produced by `_sparse_semi_structured_linear()` is of correct sizes, and `reinterpret_tensor` call again should not be needed.)

**Comment by jon-chuang** [comment]
I was able to fix the repro with a change to the meta registration, adding a transpose.

```python
    return (
        input.new_empty(
            output_sizes,
            dtype=input.dtype if input.dtype != torch.int8 else torch.int32,
        )
        .transpose(-1, -2)
        .reshape(output_sizes)
    )
```

But the original tests still fail

**Comment by jon-chuang** [comment]
Actually, this seems to have something to do with dynamic shapes. 

This strides are not correctly updated in for dynamic shapes even after a transpose!

```
INPUT STRIDE (128, 1)
OUTPUT STRIDE (s0*((128//s0)), 1)
```

**Comment by alexsamardzic** [comment]
It seems to me that `weight` input there is not of correct shape: for the test script I provided above, the shape of weight tensor is `(32, 64)`, and if I print it from within `meta_sparse_structured_linear()` method, it says `(32, 32)`.

**Comment by jon-chuang** [comment]
@alexsamardzic that's expected, as the K dimension for the matmul should be 2-4 sparse.

**Comment by jon-chuang** [comment]
@alexsamardzic this code solves the 1, 32, 64 case but doesn't solve the 1, 128, 128 case. It also seems to avoid the dynamic shapes issue. Based on what you understand about how the strides should be, can you spot any mistakes?

```python
@register_meta(aten._sparse_semi_structured_linear)
def meta_sparse_structured_linear(
    input: Tensor,
    weight: Tensor,
    _meta: Tensor,
    bias: Optional[Tensor] = None,
    _activation_opt: Optional[str] = None,
):
    output_sizes = list(input.shape)
    if bias is not None:
        assert weight.size(0) == bias.size(0), "output size mismatch"
    assert weight.size(1) == input.size(-1) / 2
    output_sizes[-1] = weight.size(0)

    transposed_strides = input.new_empty(output_sizes).transpose(-1, -2).stride()

    return input.new_empty(
        output_sizes,
        dtype=input.dtype if input.dtype != torch.int8 else torch.int32,
    ).as_strided(output_sizes, transposed_strides)
```

**Comment by jon-chuang** [comment]
@alexsamardzic I have the sinking feeling that the strides for `_sparse_semi_structured_linear` is actually some kind of complicated thing and it's not straightforward to model in meta registration.

**Comment by alexsamardzic** [comment]
Could you try with just:
```
transposed_strides = (1, input.size(0))
```

**Comment by jon-chuang** [comment]
Hmm, that worked. For both (1, 128) and (64, 128, 128). Could you explain why that works? :thinking:

**Comment by alexsamardzic** [comment]
As usual - it appears dead simple now:  Apparently, for compiled case something is doing squashing of batch dimensions before passing `input` to C++ code, and un-squashing afterwards.  So the C++ code always gets 2D `input` tensor, and it will always produce contiguous output, that just gets transposed at the end.  So the size of the output will be `(input.size(0), weight.size(0))`, but as `transpose()` is called at the end of C++ code, the strides will be exactly as above.

Should work fine for now, but it's kind of fragile: it will break if someone changes this squashing/un-squashing logic down the road, or maybe adds padding for other dimensions.  But, the test case should catch it then.

**Comment by jon-chuang** [comment]
In that case, for the meta registration, I will assert that the input/output sizes will have exactly dim 2.

**Comment by jon-chuang** [comment]
@alexsamardzic Would this be correct even in the non-squashed case:

```
    transposed_strides = (1, reduce(operator.mul, input.size()[:-1], 1))
```

**Comment by jcaip** [comment]
@alexsamardzic Thanks for the help, I forgot about the noncontiguous output. Yes, I added in 2d flattening for cusparselt, and CUTLASS uses the same code path

@jon-chuang This works because the output of _sparse_semi_structured_linear is non-contiguous as it's really the output of `(weigth @ input.T).T`. When you registered the meta registration initially, you were returning a contiguous tensor, created by new_dense. 

Now this doesn't error in the case when we don't have padding, as although the stride information is not set properly, allclose doesn't care about this and just checks that all the values are close, which they are. However, in the case when we pad the dense matrix, we additionally need to select just the non-padded values out of the resultant matrix, which we do [here](https://github.com/pytorch/pytorch/blob/b060694088bf90f66d667b274c56854c8426ee3a/torch/sparse/semi_structured.py#L355). From what I can understand this is what's reinterpret_tensor is doing, but since the strides are set differently for the tensor returned from the meta_registration and the output of _sparse_semi_structured are different we "select" the wrong line and that's why the test fails.

**Comment by jon-chuang** [comment]
Ok, I will stick to asserting on the squashed case. It is complicated and error prone to handle the strides for anything else.

**Review by jcaip — APPROVED** [review]
lgtm, thank you!

**Comment by alexsamardzic** [comment]
> @alexsamardzic Would this be correct even in the non-squashed case:
> 
> ```
>     transposed_strides = (1, reduce(operator.mul, input.size()[:-1], 1))
> ```

Not exactly, there are more strides in this case - it's not complicated to calculate them, but we'd have to differentiate between batched and non-batched case, etc.  So as we can't test it at the moment, let's indeed put an assert there.

In any case, thank you for this work!  This debugging session was actually very useful for me, namely I'm at the moment working on adding some sparse MM related features to CUTLASS, and at some point afterwards I hope to add this same thing for the CUTLASS backend of Inductor - so I would hit this stuff anyway.
